### PR TITLE
IOperation support for new pattern forms in C# 9.0.

### PR DESCRIPTION
**Pull request by gafter**

### CODE REVIEW

**Comment by 333fred** [comment]
@gafter you're missing the updated XML file. Will look when that is pushed.

**Comment by gafter** [comment]
Remaining work is tracked in [https://github.com/dotnet/roslyn/issues/41502](https://github.com/dotnet/roslyn/issues/41502)

---
In reply to: [376487045](https://github.com/dotnet/roslyn/pull/42474#issuecomment-376487045) [](ancestors = 376487045)

**Comment by gafter** [comment]
@agocke Could you please look at this?

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs [hunk 1]
```diff
@@ -1120,6 +1120,38 @@ public override void VisitConstantPattern(IConstantPatternOperation operation)
...
+        }
+
+        public override void VisitNegatedPattern(INegatedPatternOperation operation)
+        {
+            Assert.Equal(OperationKind.NegatedPattern, operation.Kind);
+            VisitPatternCommon(operation);
+            Assert.Same(operation.NegatedPattern, operation.Children.Single());
+        }
+
+        public override void VisitTypePattern(ITypePatternOperation operation)
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: >ITypePatternOperation [](start = 46, length = 21)

Consider asserting ```MatchedType``` is not null.

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs [hunk 2]
```diff
@@ -1120,6 +1120,38 @@ public override void VisitConstantPattern(IConstantPatternOperation operation)
             Assert.Same(operation.Value, operation.Children.Single());
         }
 
+        public override void VisitRelationalPattern(IRelationalPatternOperation operation)
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: >IRelationalPatternOperation [](start = 52, length = 27)

Consider asserting OperatorKind

Comment (gafter) [reply src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: Will address in [https://github.com/dotnet/roslyn/pull/42548](https://github.com/dotnet/roslyn/pull/42548)

---
In reply to: [394691115](https://github.com/dotnet/roslyn/pull/42474#discussion_r394691115) [](ancestors = 394691115)

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs (hunk 1 above)
Comment (gafter) [reply src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: Will address in [https://github.com/dotnet/roslyn/pull/42548](https://github.com/dotnet/roslyn/pull/42548)

---
In reply to: [394690849](https://github.com/dotnet/roslyn/pull/42474#discussion_r394690849) [](ancestors = 394690849)

**Pull request merged**